### PR TITLE
Add NSID plugin support for CoreDNS

### LINKED ISSUE #1256: Meta data or label support for identity/whoami?
I have a unique scenario where I would like to deploy multiple CoreDNS servers in a cluster. I want each CoreDNS server in the cluster to expose certain information like label or meta-data.

CoreDNS currently support whoami plugin that could be used to show the local IP of the CoreDNS server. Though it would be nice to allow CoreDNS to be assigned with a arbitrary label or metadata and expose it through DNS or HTTP.

For example, let's say I want to deploy 3 CoreDNS servers on 3 hosts. One host is deployed in Tokyo and another two hosts are deployed in California. I would like to have the first host with the label of:
`location=tokyo` and the rest two host with the label of: `location=california`.

A feature like the above will helps a lot in a distributed system and expand the usage of CoreDNS in more interesting ways.

### PR DESCRIPTION
**Pull request by yongtang**
### 1. What does this pull request do?

This fix adds NSID plugin support for CoreDNS, as was proposed in #1256.

### 2. Which issues (if any) are related?

This fix fixes #1256.

### 3. Which documentation changes (if any) need to be made?

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: plugin/nsid/nsid.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		}
+	}
+	return plugin.NextOrFailure(n.Name(), n.Next, ctx, w, r)
+}
+
+// WriteMsg implements the dns.ResponseWriter interface.
+func (w *ResponseWriter) WriteMsg(res *dns.Msg) error {
+	r1 := new(dns.Msg)
+
+	// Make a copy of Msg and update NSID
```
Comment (miekg) [inline plugin/nsid/nsid.go]: Why do you copy, you should be able to get away by just adding the NSID?

File: plugin/nsid/setup.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		return Nsid{Next: next, Data: nsid}
+	})
+
+	return nil
+}
+
+func nsidParse(c *caddy.Controller) (string, error) {
+	nsid := "NSID"
```
Comment (miekg) [inline plugin/nsid/setup.go]: I believe the default was the hostname (I *think*), haven't read the RFC lately

**Review by miekg — COMMENTED** [review]
Nice!
Some questions and needs (I can write it) a README.md

File: plugin/nsid/nsid.go (hunk 1 above)
Comment (yongtang) [reply plugin/nsid/nsid.go]: Thanks @miekg.

Initially I modified `e.Nsid` directly, but incorrectly not converting buffer to hex in `e.Nsid`. That results in `overflow packing` error. I though the error was caused by direct modification of `e.Nsid` and copy is needed.

Later on I fixed the hex issue, but forgot to give the direct modification a try. So the usage of copy remains.

Just tried again with direct modification and it works now. Thanks!

File: plugin/nsid/setup.go (hunk 2 above)
Comment (yongtang) [reply plugin/nsid/setup.go]: Thanks @miekg. Not familiar with this area. Now reading the RFC again, it does not mention the hostname. However, I did a google search and it turns out Bind is using hostname as the default (the related config seems to be `request-nsid`). I think it makes sense to conform to bind behavior. The PR has been updated.

**Comment by yongtang** [comment]
@miekg The PR has been updated. Please take a look.

**Review by miekg — APPROVED** [review]

**Comment by miekg** [comment]
lgtm. Here's a README
```
# nsid

*nsid* add an identifier of this server to each reply.

This plugin implements RFC 5001 and adds an EDNS0 OPT resource record to replies that uniquely
identifies the server. This can be useful in anycast setups to see which server was responsible for
generating the reply and for debugging.

## Syntax

~~~ txt
nsid [DATA]
~~~

**DATA** is the string to use in the nsid record.

If **DATA** is not given, the host's name is used.

## Examples

Enable nsid:

~~~ corefile
. {
    nsid
}
~~~
```

**Comment by yongtang** [comment]
@miekg Thanks 👍 ! Updated the PR with the README.md added.

**Pull request merged**